ipatests: nested netgroups (intg)

celestian opened the pull request:
Adds a test case for issue in SSSD that manifested in
an inability to resolve nested membership in netgroups

The test case tests for direct and indirect membership.

https://fedorahosted.org/freeipa/ticket/6439

celestian commented:
I see there are some issues:
1. It seems Travis hit some kind of time out. Maybe it is not connected to my proposed test.
2. The QuantifiedCode issues are clear.

I have PTO now and I will be back in work on Friday. But I hope I find a time to fix it sooner.

MartinBasti commented:
Travis:
```
PEP-8 errors:
./ipatests/test_integration/test_netgroup.py:87:80: E501 line too long (84 > 79 characters)
./ipatests/test_integration/test_netgroup.py:96:80: E501 line too long (88 > 79 characters)


Pylint:
Pylint is running, please wait ...
************* Module ipatests.test_integration.test_netgroup
ipatests/test_integration/test_netgroup.py:2: [W0512(invalid-encoded-data), ] Cannot decode using encoding "ascii", unexpected byte at position 9)
ipatests/test_integration/test_netgroup.py:87: [E1101(no-member), TestNetgroups.check_users_in_netgroups] Class 'domain' has no 'name' member)
ipatests/test_integration/test_netgroup.py:96: [E1101(no-member), TestNetgroups.check_nested_netgroup_hierarchy] Class 'domain' has no 'name' member)
ipatests/test_integration/test_netgroup.py:125: [E1101(no-member), TestNetgroups.test_remove_nested_netgroup] Class 'domain' has no 'name' member)
ipatests/test_integration/test_netgroup.py:130: [E1101(no-member), TestNetgroups.test_remove_nested_netgroup] Class 'domain' has no 'name' member)
ipatests/test_integration/test_netgroup.py:131: [E1101(no-member), TestNetgroups.test_remove_nested_netgroup] Class 'domain' has no 'name' member)
ipatests/test_integration/test_netgroup.py:132: [E1101(no-member), TestNetgroups.test_remove_nested_netgroup] Class 'domain' has no 'name' member)
ipatests/test_integration/test_netgroup.py:147: [E1101(no-member), TestNetgroups.test_remove_nested_netgroup] Class 'domain' has no 'name' member)
ipatests/test_integration/test_netgroup.py:148: [E1101(no-member), TestNetgroups.test_remove_nested_netgroup] Class 'domain' has no 'name' member)
ipatests/test_integration/test_netgroup.py:149: [E1101(no-member), TestNetgroups.test_remove_nested_netgroup] Class 'domain' has no 'name' member)
make: *** [pylint] Error 6
```

I'm afraid that pylint cannot handle it, so probably you have to disable problematic lines.

MartinBasti commented on ipatests/test_integration/test_netgroup.py:
```diff
@@ -0,0 +1,149 @@
+                                        '--users=testuser%d' % i,
+                                        'test_netgroup%d' % i])
+
+    def teardown_three_netgroups():
```
use test_data from previous comment too here please

MartinBasti replied:
and it can be in one cycle

MartinBasti commented on ipatests/test_integration/test_netgroup.py:
```diff
@@ -0,0 +1,149 @@
+
+    def teardown_three_netgroups():
+        for i in range(1, 4):
+            request.cls.master.run_command(['ipa', 'netgroup-del',
```
You need to prevent failures in teardown method.

Please use `run_command(['*'], raiseonerr=False)` it will not raise an exception when failure occurs

MartinBasti commented on ipatests/test_integration/test_netgroup.py:
```diff
@@ -0,0 +1,149 @@
+        cls.domain = cls.get_domains()[0]
+
+    @classmethod
+    def uninstall(cls, mh):
```
You don't need to explicitly add `uninstall` method if you call only `super()` inside it is automatic.

MartinBasti commented on ipatests/test_integration/test_netgroup.py:
```diff
@@ -0,0 +1,149 @@
+        super(TestNetgroups, cls).install(mh)
+
+        cls.client = cls.clients[0]
+        cls.clientname = cls.client.run_command(
```
Why do you need a `clientname` AFAIK client should contain hostname as attribute.

MartinBasti commented on ipatests/test_integration/test_netgroup.py:
```diff
@@ -0,0 +1,149 @@
+        cls.clientname = cls.client.run_command(
+            ['hostname', '-s']).stdout_text.strip()
+
+        cls.domain = cls.get_domains()[0]
```
IMHO cls.domain should already contain the domain

MartinBasti replied:
If you remove all cls.<attr> definitions in install method, you can remove it

MartinBasti commented on ipatests/test_integration/test_netgroup.py:
```diff
@@ -0,0 +1,149 @@
+    def install(cls, mh):
+        super(TestNetgroups, cls).install(mh)
+
+        cls.client = cls.clients[0]
```
Could you rather use 
```
client = self.clients[0]
```

directly in tests? It is more consistent with other tests

MartinBasti commented on ipatests/test_integration/test_netgroup.py:
```diff
@@ -0,0 +1,149 @@
+    def check_users_in_netgroups(self):
+        clear_sssd_cache(self.client)
+        for i in range(1, 4):
+            result = self.client.run_command(['getent', 'passwd',
```
you have to add `raiseonerr=False` to receive result, otherwise exception will be raised,  or you can remove `assert result.returncode == 0`

MartinBasti commented on ipatests/test_integration/test_netgroup.py:
```diff
@@ -0,0 +1,149 @@
+
+    def check_nested_netgroup_hierarchy(self):
+        clear_sssd_cache(self.client)
+        for i in range(1, 4):
```
IMO nesting hierarchy can be represent similarly as test_data

celestian commented:
I addressed all comment, I hope.
I know that there are still some pylint stuff like:
```
ipatests/test_integration/test_netgroup.py:91: [E1101(no-member), TestNetgroups.check_users_in_netgroups] Class 'domain' has no 'name' member)
ipatests/test_integration/test_netgroup.py:108: [E1101(no-member), TestNetgroups.check_nested_netgroup_hierarchy] Class 'domain' has no 'name' member)
ipatests/test_integration/test_netgroup.py:132: [E1101(no-member), TestNetgroups.test_remove_nested_netgroup] Class 'domain' has no 'name' member)
```

I don't know why due to line 228 and 230 in pylint_plugins.py. Missed I something?

celestian commented:
I addressed issue in Travis and Quantified.
New version pushed.

PS: Right way is ```self.master.domain.name```

MartinBasti commented on ipatests/test_integration/test_netgroup.py:
```diff
@@ -0,0 +1,183 @@
+#
+# Copyright (C) 2015  FreeIPA Contributors see COPYING for license
```
2017 please

celestian commented:
2015-->2017 addressed. New version pushed.

MartinBasti commented on ipatests/test_integration/test_netgroup.py:
```diff
@@ -0,0 +1,183 @@
+        }
+        test_data.append(data)
+
+        members = [d['user']['login'] for d in test_data] if test_data else []
```
You dont need,  `if test_data else []` it behaves the same when test_data == []

MartinBasti commented on ipatests/test_integration/test_netgroup.py:
```diff
@@ -0,0 +1,183 @@
+                'last': 'User_{}'.format(i),
+            },
+            'netgroup': 'testgroup_{}'.format(i),
+            'nested_netgroup': 'testgroup_{}'.format(i-1) if i > 1 else ''
```
I prefer `None` for nothing, instead of empty string

MartinBasti commented on ipatests/test_integration/test_netgroup.py:
```diff
@@ -0,0 +1,183 @@
+from ipatests.test_integration.tasks import clear_sssd_cache
+
+
+def get_test_data():
```
Could be test_data a module variable instead of function, I don't see reason to generate the same data again and again

MartinBasti commented on ipatests/test_integration/test_netgroup.py:
```diff
@@ -0,0 +1,183 @@
+    Test Netgroups
+    """
+
+    num_clients = 1
```
Can be this tested on server side? Do we need to install extra machine?

celestian commented:
Addressed, pushed.

rcritten commented on Contributors.txt:
```diff
@@ -98,6 +98,7 @@ Developers:
 	Adam Young
 	Jan Zelený
 	Pavel Zůna
+	Petr Čech
```
nit: this file is sorted alphabetically

celestian replied:
Addressed, thanks.

MartinBasti commented:
Fixed upstream
master:
https://fedorahosted.org/freeipa/changeset/dc99d3c04e43b08d2364209a641b8b9111e5986c
